design/baremetalhost-v1alpha2.md

```diff
@@ -0,0 +1,223 @@
+format, however, because we want to be able to drop support
+quickly. Therefore, we should not write a conversion web hook.
+
+Instead, we should add startup code to the baremetal-operator to read
+all host resources in their current format and write them (with type
+conversion but no other changes) to the new storage format. See option
+2
+[here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version)
+for more discussion.
+
+### Dropping Hardware Profile Field
+
+Now that we have support for more verbose root device hints and we
+have the `hardware-classification-controller`, we no longer need the
+hardware profile fields. We can remove them as part of the
+backwards-incompatible changes as we move to the new API version.
+
+### Risks and Mitigations
+
+We do not know all of the ways that consumers of the API are using the
+CPU speed values currently being reported. Given that most kubernetes
+tools appear to assume that floating point types are not to be used in
+APIs, it is likely that most consumers are either ignoring the value
+or using it in simple ways (displaying it, logging it, etc.) where
+loss of precision from the type conversion to a float is not
+significant.
+
+### Work Items
+
+- Add new `v1alpha2` API
+- Change the `ClockSpeed` type to `string` in `v1alpha2`.
+- Remove hardware profile from spec and status.
+- Remove the state machine states for setting the hardware profile.
+- Set `v1alpha2` as the storage version
+- Add conversion code to the `baremetal-operator`
+- Update the inspection code to no longer convert the value given by
+  Ironic Inspector before storing it.
+- Release a version of the operator that can upgrade existing
+  resources.
+- Drop support for `v1alpha1`
+  - Remove the API code
+  - Remove the conversion code
+  - Remove the `operator-sdk` code from the `tools` directory
+- Release the operator.
+
+### Dependencies
+
+The prerequisite work was done as part of
+[#434](https://github.com/metal3-io/baremetal-operator/issues/434).
+
+### Test Plan
+
+- Test updating the `baremetal-operator` on an existing cluster with
+  `v1alpha1` host resources.
+- Unit tests, etc.
+
+### Upgrade / Downgrade Strategy
+
+This approach supports upgrading, but not downgrading.
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks
+
+None
+
+## Alternatives
+
+### Change Type in Place, Retain v1alpha1
+
+The API version for `BareMetalHost` is `v1alpha1`. Upstream in metal3,
+we could reasonably say that the way to fix the problem with the CPU
+speed field is to change the type of the field and that no backwards
+compatibility or upgrade support is provided. Fixing the type in place
+allows the `operator-sdk` tools to work properly again.
+
+Any existing users would be broken by the type change, because
+existing data could no longer be deserialized from storage properly.
```

Comment:
Have we verified that this is actually the case? All numbers in JSON are stored as floats anyway.

Reply:
If we change from float to string, I think they're going to be broken. If we change to int, maybe not all of them will? I would expect any golang consumers to break either way. Other languages may be more flexible.